pennylane/templates/subroutines/qmc.py

```diff
@@ -0,0 +1,358 @@
+    case,
+
+    .. math::
+
+        N =  \mathcal{O}\left(\frac{1}{\epsilon^{2}}\right).
+
+    Hence, the quantum Monte Carlo algorithm has a quadratically improved time complexity with
+    :math:`N`.
+
+    Args:
+        probs (array): input probability distribution as a flat array
+        func (callable): input random variable defined on the set of integers
+            :math:`X = [0, 1, \ldots, M - 1]` with output value inside :math:`[0, 1]`
+        target_wires (Union[Wires, Sequence[int], or int]): the target wires
+        estimation_wires (Union[Wires, Sequence[int], or int]): the estimation wires
+
+    Raises:
+        ValueError: if ``probs`` is not flat or has a length that is not compatible with
+            ``target_wires``
+
+    .. note::
+
+        This template is only compatible with simulators because the algorithm is performed using
+        unitary matrices. To implement the quantum Monte Carlo algorithm on hardware requires
+        breaking down the unitary matrices into hardware-compatible gates.
+
+    .. UsageDetails::
+
+        Consider a standard normal distribution :math:`p(x)` and a random variable
+        :math:`f(x) = \sin ^{2} (x)`. The expectation value of :math:`f(x)` is
+        :math:`\int_{-\infty}^{\infty}f(x)p(x) \approx 0.432332`. This number can be approximated by
+        discretizing the problem and using the quantum Monte Carlo algorithm.
+
+        First, the problem is discretized:
+
+        .. code-block:: python
+
+            from scipy.stats import norm
+
+            m = 5
+            M = 2 ** m
+
+            xmax = np.pi  # bound to region [-pi, pi]
+            xs = np.linspace(-xmax, xmax, M)
+
+            probs = np.array([norm().pdf(x) for x in xs])
+            probs /= np.sum(probs)
+
+            func = lambda i: np.sin(xs[i]) ** 2
+
+        The ``QuantumMonteCarlo`` template can then be used:
+
+        .. code-block::
+
+            n = 10
+            N = 2 ** n
+
+            target_wires = range(m + 1)
+            estimation_wires = range(m + 1, n + m + 1)
+
+            dev = qml.device("default.qubit", wires=(n + m + 1))
+
+            @qml.qnode(dev)
+            def circuit():
+                qml.templates.QuantumMonteCarlo(
+                    probs,
+                    func,
+                    target_wires=target_wires,
+                    estimation_wires=estimation_wires,
+                )
+                return qml.probs(estimation_wires)
+
+            phase_estimated = np.argmax(circuit()[:int(N / 2)]) / N
+
+        The estimated value can be read-out as
+
+        >>> (1 - np.cos(np.pi * phase_estimated)) / 2
+        0.4327096457464369
+    """
+    if isinstance(probs, np.ndarray) and probs.ndim != 1:
+        raise ValueError("The probability distribution must be specified as a flat array")
+
+    dim_p = len(probs)
+    num_target_wires = float(np.log2(2 * dim_p))
+
+    if not num_target_wires.is_integer():
```

Comment:
I thought `ceil(log2(dim_p))` would be enough qubits to encode dim_p probabilities? Or why do we multiply by 2?

Reply:
@ixfoduap I think this could cause problems if users do:
```python
m = 5
M = 2 ** m - 3  # (2 ** 5 = 32)

xmax = np.pi  # bound to region [-pi, pi]
xs = np.linspace(-xmax, xmax, M)

probs = np.array([norm().pdf(x) for x in xs])
probs /= np.sum(probs)

func = lambda i: np.sin(xs[i]) ** 2
```
Their function is only defined on the set of `xs`, which has length `29`. If we pad the probability distribution to `32`, then the QMC function will try to evaluate `func(31)`, which is not defined.

I like the idea, but prefer not to do too many implicit things. For example, this is something a user could do without too much overhead.

Reply:
@DSGuala yes the x2 is for the ancilla.

Reply:
I see what you mean. I still find it a bit strange to restrict to powers of 2, but I guess in a sense this is a requirement of how the algorithm works. I'm fine with leaving as is, hopefully won't cause much trouble to users